### PR TITLE
Use operand bundles for gc-transition arguments

### PR DESCRIPTION
**Pull request by JosephTremoulet**
The prior approach of eagerly generating gc.statepoint calls with
gc-transition arguments is being deprecated (see e.g. r258551), in favor
of attaching these as operand bundles, which should make the semantics
less opaque during optimization.

Since the lowering code that looks for the bundles is in
RewriteStatepointsForGC, we now need to run that pass even with
conservative GC, in methods that have gc transition bundles.  Add a
post-pass to mark all other calls "gc-leaf-function" in that configuration
so that RewriteStatepointsForGC will leave them alone.  Expose a getter on
GenIR to indicate whether we had to generate a managed frame so that we
can avoid this in the common case.

Regardless of GC flavor, inline asm calls (which we use for nops) now must
be explicitly annotated "gc-leaf-function" to avoid lowering trying to
process them as statepoints.

### CODE REVIEW

**Comment by JosephTremoulet** [comment]
@pgavlin / @swaroop-sridhar PTAL

File: include/Reader/readerir.h [hunk 1]
```diff
@@ -814,6 +814,12 @@ class GenIR : public ReaderBase {
                          bool IsRelocatable, bool IsCallTarget,
                          bool IsFrozenObject = false);
 
+  /// Indicate whether this method needed an unmanaged frame generated for it
+  /// (to make a call to unmanaged code).
+  ///
+  /// \returns True iff an unmanaged frame was created.
+  bool getHasUnmanagedFrame() { return UnmanagedCallFrame != nullptr; }
```
Comment (pgavlin) [inline include/Reader/readerir.h]: Nit: I might call this `containsUnmanagedCall()` for the sake of clarity. Similar feedback applies for the output parameter added above.

**Comment by pgavlin** [comment]
LGTM aside from one nit. This is much cleaner than eagerly emitting statepoints; thanks for making the change.

**Comment by swaroop-sridhar** [comment]
Green for me too, thanks.

**Comment by JosephTremoulet** [comment]
Updated with the suggested rename.  Will merge if the original CI testing (http://dotnet-ci.cloudapp.net/view/dotnet_llilc/job/dotnet_llilc_prtest/1406/) finishes cleanly.

**Pull request merged**